documents/coredocument.go

```diff
@@ -231,53 +240,81 @@ func (cd *CoreDocument) addNewRule(role *coredocumentpb.Role, action coredocumen
 	cd.Document.ReadRules = append(cd.Document.ReadRules, rule)
 }
 
+// TreeProof is a helper structure to pass to create proofs
+type TreeProof struct {
+	tree       *proofs.DocumentTree
+	treeHashes [][]byte
+}
+
+// newTreeProof returns a TreeProof instance pointer
+func newTreeProof(t *proofs.DocumentTree, th [][]byte) *TreeProof {
+	return &TreeProof{tree: t, treeHashes: th}
+}
+
 // CreateProofs takes Document data tree and list to fields and generates proofs.
 // we will try generating proofs from the dataTree. If failed, we will generate proofs from CoreDocument.
 // errors out when the proof generation is failed on core Document tree.
-func (cd *CoreDocument) CreateProofs(docType string, dataTree *proofs.DocumentTree, fields []string) (proofs []*proofspb.Proof, err error) {
-	srpHashes, err := cd.GetSigningRootProof()
+func (cd *CoreDocument) CreateProofs(docType string, dataTree *proofs.DocumentTree, fields []string) (prfs []*proofspb.Proof, err error) {
+	treeProofs := make(map[string]*TreeProof, 3)
+
+	signatureTree, err := cd.getSignatureDataTree()
 	if err != nil {
-		return nil, errors.New("failed to generate signing root proofs: %v", err)
+		return nil, errors.New("failed to generate signatures tree: %v", err)
 	}
-
 	cdTree, err := cd.documentTree(docType)
 	if err != nil {
 		return nil, errors.New("failed to generate core Document tree: %v", err)
 	}
+	srpHashes, err := cd.GetSigningRootProof()
+	if err != nil {
+		return nil, errors.New("failed to generate signing root proofs: %v", err)
+	}
 
 	dataRoot := dataTree.RootHash()
 	cdRoot := cdTree.RootHash()
 
-	// try generating proofs from data root first
-	proofs, missedPfs := generateProofs(dataTree, fields, append([][]byte{cdRoot}, srpHashes...))
-	if len(missedPfs) == 0 {
-		return proofs, nil
+	dataPrefix, err := getDataTreePrefix(dataTree)
+	if err != nil {
+		return nil, err
 	}
 
-	// generate proofs from cdTree. fail if any proofs are missed after this
-	pfs, missedPfs := generateProofs(cdTree, missedPfs, append([][]byte{dataRoot}, srpHashes...))
-	if len(missedPfs) > 0 {
-		return nil, errors.New("failed to generate proofs for %v", missedPfs)
-	}
+	treeProofs[dataPrefix] = newTreeProof(dataTree, append([][]byte{cdRoot}, srpHashes...))
+	treeProofs[SignaturesTreePrefix] = newTreeProof(signatureTree, srpHashes)
+	treeProofs[CDTreePrefix] = newTreeProof(cdTree, append([][]byte{dataRoot}, srpHashes...))
+
+	return generateProofs(fields, treeProofs)
+}
 
-	proofs = append(proofs, pfs...)
-	return proofs, nil
+func getDataTreePrefix(dataTree *proofs.DocumentTree) (string, error) {
```

Comment:
This can be removed as soon as we have merged: https://github.com/centrifuge/precise-proofs/pull/72.
And to merge that we have to integrate latest precise proofs changes as currently they break go-centrifuge.